sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java

```diff
@@ -322,6 +322,36 @@ private JobService getJobService(PipelineOptions pipelineOptions) throws IOExcep
       return jobService;
     }
 
+    @Teardown
+    public void onTeardown() {
+      try {
+        if (datasetService != null) {
+          datasetService.close();
+          datasetService = null;
+        }
+        if (jobService != null) {
+          jobService.close();
+          jobService = null;
+        }
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private DatasetService getDatasetService(PipelineOptions pipelineOptions) throws IOException {
+      if (datasetService == null) {
+        datasetService = bqServices.getDatasetService(pipelineOptions.as(BigQueryOptions.class));
+      }
+      return datasetService;
+    }
+
+    private JobService getJobService(PipelineOptions pipelineOptions) throws IOException {
+      if (jobService == null) {
+        jobService = bqServices.getJobService(pipelineOptions.as(BigQueryOptions.class));
+      }
+      return jobService;
+    }
+
```

Comment:
Also this

Reply:
I made this change against the v2.31.0 tag, then had to merge-in HEAD for Github to allow PR creation. Only once I'd pushed the button did I realise I'd captured irrelevant commits. 
Really, really, sorry. What do you recommend?

Reply:
Files reverted.